File: RomanToInteger.java
// Roman numerals are usually written largest to smallest from left to right.
// However, the numeral for four is not IIII. Instead, the number four is written as IV.
// Because the one is before the five we subtract it making four. The same principle applies to the number nine,
// which is written as IX. There are six instances where subtraction is used

// Given a roman numeral, convert it to an integer.

import java.util.HashMap;

public class RomanToInteger {
    public int romanNumeralToInt(String s) {
        // create a hashmap to store the values of the roman numerals
        HashMap<Character, Integer> values = new HashMap<>();
        values.put('I', 1);
        values.put('V', 5);
        values.put('X', 10);
        values.put('L', 50);
        values.put('C', 100);
        values.put('D', 500);
        values.put('M', 1000);

        int result = 0;

        // Iterate through String s or the Roman Numeral input string
        for (int i = 0; i < s.length(); i++) {
            int currentValue = values.get(s.charAt(i));

            // If the next symbol is larger, subtract the current symbols value
            if (i < s.length() - 1 && currentValue < values.get(s.charAt(i + 1))) {
                result -= currentValue;
            } else {
                result += currentValue;
            }
        }
        return result;
    }
}

// Create a hashmap to store the kay value pairs that we created to represent the roman numeral letter and
// its value
// the for loop iterates, then the .get method gets the VALUE (integer) part of the kay value pair
// The if statement checks if the end of the string has been reached, then it checks the order of the
// letters and sees if subtraction is necessary for example IV - the lower value is infront of the higher

File: RomanToInteger.py
# Roman numerals are usually written largest to smallest from left to right.
# However, the numeral for four is not IIII. Instead, the number four is written as IV.
# Because the one is before the five we subtract it making four. The same principle applies to the number nine,
# which is written as IX. There are six instances where subtraction is used

# Given a roman numeral, convert it to an integer.

# String manipulation
class Solution():
    def romanToInt(self, s):

        # create dict to store values to roman lettering
        roman_num_dict = {
            "I": 1, "V": 5, "X": 10, "L": 50, "C": 100, "D": 500, "M": 1000
        }

        total = 0
        previous_value = 0

        for char in reversed(s): # REVERSED allows us to iterate through the string BACKWARDS. can be useful!
            current_value = roman_num_dict[char]

            # if the current value is less than the previous one, we will have to subtract it. ex: IX will be 9 and the right-most letter equates to a 10
            if current_value < previous_value:
                total = total - current_value

            else:
                total = total + current_value
            
            previous_value = current_value # this updates the previous value for the next iteration
        
        return total
    
my_solution = Solution()

answer = my_solution.romanToInt("IL")

print(answer)
